### PR TITLE
hypre default LAI

### PR DESCRIPTION
**Pull request by castelletto1**
This PR sets the default linear algebra interface (LAI) to _hypre_. In particular all integrated tests are now using UMFPACK (serial direct) as the linear solver. The following integrated tests are temporarily disabled:

- tests relying on the `HydrofractureSolver` (monolithic system matrix is currently missing)
  1. `Hydrofracture_KGD_NodeBased_C3D6_01`, `Hydrofracture_KGD_NodeBased_C3D6_09`
  1. `KGD_ZeroToughness_01`, `KGD_ZeroToughness_06`
  1. `KGD_ZeroViscosity_01`, `KGD_ZeroViscosity_06`

- `ContactMechanics_PassingCrack` (this test will be reintroduced with PR #986)

Note that for integrated tests `PoroElastic_Mandel` and `beamBending` a looser tolerance to check the direct solver solution was specified in the xml input file (`checkResidualTolerance`) because of the conditioning of the system matrix. PR #1169 will remove the need for explicitly set such value by computing it using an estimate of the condition number of the system matrix.

Additional changes introduced in this PR:
- treatment of the zero right-hand side case for _hypre_, PETSc and Trilinos
- sub-communicator strategy for handling in Suite Sparse interface cases in which the system matrix has no rows living on process 0 (e.g., integrated test `fractureMatrixFlow_2d_04`)

Requires:
- [ ] GEOSX/integratedTests#94

### CODE REVIEW

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.hpp [hunk 1]
```diff
@@ -52,6 +52,7 @@ struct SuiteSparseData
   real64 Control[UMFPACK_CONTROL];    //!< SuiteSparse options
   void * Symbolic;                    //!< pointer to the symbolic factorization
   void * Numeric;                     //!< pointer to the numeric factorization
+  MPI_Comm globalComm;                //!< MPI communicator
   MPI_Comm comm;                      //!< MPI communicator
```
Comment (af1990) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.hpp:56]: Perhaps an appropriate comment to distinguish the two communicators would be nice.

File: src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_1d.xml [hunk 2]
```diff
@@ -16,7 +16,9 @@
         newtonTol="1.0e-6"
         newtonMaxIter="15"/>
       <LinearSolverParameters
-        solverType="direct"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_1d.xml:20]: Indentation issue?

File: src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_cap_1d.xml [hunk 3]
```diff
@@ -17,7 +17,9 @@
         newtonTol="1.0e-6"
         newtonMaxIter="15"/>
       <LinearSolverParameters
-        solverType="direct"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_cap_1d.xml:21]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_PassingCrack.xml [hunk 4]
```diff
@@ -20,6 +20,7 @@
         maxTimeStepCuts="2"/>
       <LinearSolverParameters
         solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_PassingCrack.xml:23]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_SimpleCubes.xml [hunk 5]
```diff
@@ -20,6 +20,7 @@
         maxTimeStepCuts="2"/>
       <LinearSolverParameters
         solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_SimpleCubes.xml:23]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_UnstructuredCrack.xml [hunk 6]
```diff
@@ -20,6 +20,7 @@
         maxTimeStepCuts="2"/>
       <LinearSolverParameters
         solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/ContactMechanics_UnstructuredCrack.xml:23]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/FlowProppantBedTransport_2d.xml [hunk 7]
```diff
@@ -20,8 +20,9 @@
         newtonMinIter="1"
         maxTimeStepCuts="5"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-7"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/FlowProppantBedTransport_2d.xml:24]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/FlowProppantTransport_2d.xml [hunk 8]
```diff
@@ -19,8 +19,9 @@
         lineSearchAction="None"
         newtonMinIter="1"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-12"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/FlowProppantTransport_2d.xml:23]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/PoroElastic_Mandel_FIM.xml [hunk 9]
```diff
@@ -17,9 +17,9 @@
         maxTimeStepCuts="1"
         lineSearchMaxCuts="0"/>
       <LinearSolverParameters
-        solverType="gmres"
-        preconditionerType="block"
-        krylovMaxIter="200"
+        solverType="direct"
+        directCheckResTol="1.e-9"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/PoroElastic_Mandel_FIM.xml:22]: Indentation issue?

File: src/coreComponents/physicsSolvers/multiphysics/integratedTests/PoroElastic_Mandel_FIM.xml [hunk 10]
```diff
@@ -17,9 +17,9 @@
         maxTimeStepCuts="1"
         lineSearchMaxCuts="0"/>
       <LinearSolverParameters
-        solverType="gmres"
-        preconditionerType="block"
-        krylovMaxIter="200"
+        solverType="direct"
+        directCheckResTol="1.e-9"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/multiphysics/integratedTests/PoroElastic_Mandel_FIM.xml:21]: This should be fixed if PR #1169 is approved. To be checked later ...

File: src/coreComponents/physicsSolvers/simplePDE/integratedTests/10x10x10_LaplaceFEM.xml [hunk 11]
```diff
@@ -8,7 +8,12 @@
       timeIntegrationOption="SteadyState"
       fieldName="Temperature"
       logLevel="0"
-      targetRegions="{ Region1 }"/>
+      targetRegions="{ Region1 }">
+      <LinearSolverParameters
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/simplePDE/integratedTests/10x10x10_LaplaceFEM.xml:14]: Indentation issue?

File: src/coreComponents/physicsSolvers/simplePDE/integratedTests/50x10x5_LaplaceFEM.xml [hunk 12]
```diff
@@ -8,7 +8,12 @@
       timeIntegrationOption="SteadyState"
       fieldName="Temperature"
       logLevel="0"
-      targetRegions="{ Region1 }"/>
+      targetRegions="{ Region1 }">
+      <LinearSolverParameters
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/simplePDE/integratedTests/50x10x5_LaplaceFEM.xml:14]: Indentation issue?

File: src/coreComponents/physicsSolvers/solidMechanics/integratedTests/SSLE-QS-beamBending.xml [hunk 13]
```diff
@@ -18,9 +18,10 @@
         newtonTol="1.0e-6"
         newtonMaxIter="8"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-12"
-        preconditionerType="amg"/>
+        solverType="direct"
+        directParallel="0"
+	directCheckResTol="1.e-10"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/solidMechanics/integratedTests/SSLE-QS-beamBending.xml:23]: Indentation issue?
As before, with PR #1169 this should be avoided ...

File: src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_StaticPenny_PrismElem.xml [hunk 14]
```diff
@@ -14,8 +14,9 @@
         newtonTol="1.0e-3"
         newtonMaxIter="8"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-10"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_StaticPenny_PrismElem.xml:18]: Indentation issue?

File: src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_ThreeNodesPinched_SlantFrac.xml [hunk 15]
```diff
@@ -14,8 +14,9 @@
         newtonTol="1.0e-3"
         newtonMaxIter="8"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-10"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_ThreeNodesPinched_SlantFrac.xml:18]: Indentation issue?

File: src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_ThreeNodesPinched_HorizontalFrac.xml [hunk 16]
```diff
@@ -14,8 +14,9 @@
         newtonTol="1.0e-3"
         newtonMaxIter="8"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-10"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/DryFrac_ThreeNodesPinched_HorizontalFrac.xml:18]: Indentation issue?

File: src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/SurfaceGenerator.xml [hunk 17]
```diff
@@ -14,8 +14,9 @@
         newtonTol="1.0e-3"
         newtonMaxIter="8"/>
       <LinearSolverParameters
-        solverType="gmres"
-        krylovTol="1.0e-10"/>
+        solverType="direct"
+	directParallel="0"
```
Comment (af1990) [inline src/coreComponents/physicsSolvers/surfaceGeneration/integratedTests/SurfaceGenerator.xml:18]: Indentation issue?

**Review by af1990 — COMMENTED** [review]
Well done @castelletto1!
Right now I am trying to run all the integratedTests.

File: src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_1d.xml (hunk 2 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/fluidFlow/integratedTests/compositionalMultiphaseFlow/4comp_2ph_1d.xml:20]: OK, will fix this everywhere.

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.cpp [hunk 18]
```diff
@@ -229,68 +234,80 @@ int SuiteSparseSolve( SuiteSparseData & SSData,
...
-    delete [] SSData.data;
-    GEOSX_LAI_CHECK_ERROR( hypre_CSRMatrixDestroy( SSData.CSRmatrix ) );
+    int const rank = MpiWrapper::Comm_rank( SSData.comm );
+    if( rank == SSData.workingRank )
+    {
+      umfpack_dl_free_symbolic( &SSData.Symbolic );
+      umfpack_dl_free_numeric( &SSData.Numeric );
+      delete [] SSData.rowPtr;
+      delete [] SSData.colIndices;
+      delete [] SSData.data;
```
Comment (klevzoff) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.cpp:307]: We should make these `vector` or `array1d`

**Review by klevzoff — APPROVED** [review]
My comment is not on the changes in this PR, just something I missed when reviewing direct solver PR. 

This usage pattern
```c++
SuiteSparseData SSData;
SuiteSparseCreate( mat, parameters, SSData );
SuiteSparseSetup( SSData, timeSetup );
SuiteSparseSolve( SSData, rhs, sol, timeSolve );
SuiteSparseDestroy( SSData );
``` 
suggests that `SuiteSparseData` should be a proper c++ class (same for `SuperLU_DistData`). This will be especially important if we start using exceptions to propagate errors later on (instead of just aborting the program). Let's refactor it in a separate PR?

**Comment by af1990** [comment]
> This usage pattern
> 
> ```c++
> SuiteSparseData SSData;
> SuiteSparseCreate( mat, parameters, SSData );
> SuiteSparseSetup( SSData, timeSetup );
> SuiteSparseSolve( SSData, rhs, sol, timeSolve );
> SuiteSparseDestroy( SSData );
> ```
> 
> suggests that `SuiteSparseData` should be a proper c++ class (same for `SuperLU_DistData`).

You're right. This was my first implementation, but then I simplify everything, being just an interface to SuperLU/UMFPACK hidden behind the _hypre_ interface. But since there's a valid reason, we can adopt the class design.

I think that doing this in another PR is fine.
@castelletto1?

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.hpp (hunk 1 above)
Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.hpp]: Done.

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.cpp (hunk 18 above)
Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuiteSparse.cpp:307]: Agreed. Let's tackle everything in a separate PR.

**Comment by castelletto1** [comment]
> I think that doing this in another PR is fine.
> @castelletto1?

Agreed. I need to rebaseline all tests again after #1079 got merged. Then, we should be good to go.

**Pull request closed without merging**